authoringAssistant/src/main/java/explorableviz/transparenttext/QueryContext.java

```diff
@@ -220,4 +256,39 @@ public static ArrayList<QueryContext> loadCases(String casesFolder, int numInsta
                 });
         return queryContexts;
     }
+
+    public void replaceVariables(Map<String, String> variables, Random random) {
```

Comment:
I'm wondering why we need to support mutable `QueryContext`s, where the object is first constructed and then later mutated. Would it not make more sense for the `replaceVariables` behaviour to be part of the initialisation of the `QueryContext`? Could they then be immutable?

Reply:
For example, suppose `replaceVariables` (instead of being hardcoded to modify specific fields in the `QueryContext`) instead just took an input string and transformed it using `variables` and `random`. Couldn't we then just call `replaceVariables` as we set the initial value of `code`, `expected`, `loadedDatasets`, etc?

Reply:
I got the point, and i did some refactoring on the code to separate the concept of `QueryContext `with the concept of `TextQueryContext`.